src/FieldMapB.cpp

```diff
@@ -0,0 +1,365 @@
+    int ny = int((maxs[1] - mins[1]) / steps[1]) + 2;
+    int nz = int((maxs[2] - mins[2]) / steps[2]) + 2;
+
+    Bvals_XYZ.resize(nx);
+    for (auto& B3 : Bvals_XYZ) {
+      B3.resize(ny);
+      for (auto& B2 : B3) {
+        B2.resize(nz);
+        for (auto& B : B2) {
+          B.resize(3, 0.); // Bx, By, Bz
+        }
+      }
+    }
+  }
+}
+
+// get cell indices corresponding to point of interest
+void FieldMapB::GetIndices(std::vector<double> coord, std::vector<int> *indices, std::vector<double> *deltas)
+{
+  for( uint i = 0; i < coord.size(); i++ ) {
+     // Range check
+     if( coord[i] < mins[i] || coord[i] > maxs[i] ) {
+       indices->clear();
+       indices->resize( coord.size(), -1 );
+       return;
+     }
+
+     double temp_idx;
+     deltas->push_back( std::modf( (coord[i] - mins[i]) / steps[i], &temp_idx ) );
+     indices->push_back( static_cast<int>(temp_idx) );
+  }
+}
+
+// load data
+void FieldMapB::LoadMap(const std::string& map_file, double scale)
+{
+  std::string   line;
+  std::ifstream input(map_file);
+  if (!input) {
+    std::cout << "FieldMapB Error: file \"" << map_file << "\" cannot be read." << std::endl;
+  }
+
+  std::vector<double> coord = {};
+  std::vector<double> Bcomp = {};
+  std::vector<int> indices = {};
+  std::vector<double> deltas = {};
+
+  while (std::getline(input, line).good()) {
+    std::istringstream iss(line);
+
+    coord.clear();
+    Bcomp.clear();
+    indices.clear();
+    deltas.clear();
+
+    if( coord_type.compare("BrBz") == 0 ) {
+      coord.resize(2);
+      Bcomp.resize(2);
+      iss >> coord[0] >> coord[1] >> Bcomp[0] >> Bcomp[1];
+    }
+    if( coord_type.compare("BxByBz") == 0 ) {
+      coord.resize(3);
+      Bcomp.resize(3);
+      iss >> coord[0] >> coord[1] >> coord[2] >> Bcomp[0] >> Bcomp[1] >> Bcomp[2];
+    }
+
+    GetIndices(coord, &indices, &deltas);
+
+    if (std::count( indices.begin(), indices.end(), -1)) {
+      std::cout << "FieldMapB Warning: coordinates out of range, skipped it."
+        << std::endl;
+    } else {
+      if( coord_type.compare("BrBz") == 0 ) {
+        Bvals_RZ[ indices[0] ][ indices[1] ] = {Bcomp[0] * scale, Bcomp[1] * scale};
+        //ROOT::Math::XYZPoint p(coord[0], 0, coord[1]);
+        //std::cout << p << " -> " << trans*p << std::endl;
+        //std::cout << indices[0] << ", " << indices[1] << ", " << Bcomp[0] << ", " << Bcomp[1] << std::endl;
+      }
+      if( coord_type.compare("BxByBz") == 0 ) {
+        Bvals_XYZ[ indices[0] ][ indices[1] ][ indices[2] ] = {Bcomp[0] * scale, Bcomp[1] * scale, Bcomp[2] * scale};
+      }
+    }
+  }
+}
+
+// get field components
+void FieldMapB::fieldComponents(const double* pos, double* field)
+{
+  // coordinate conversion
+  auto p = trans_inv * ROOT::Math::XYZPoint(pos[0], pos[1], pos[2]);
+
+  std::vector<int>  indices = {};
+  std::vector<double> deltas = {};
+  ROOT::Math::XYZPoint B = ROOT::Math::XYZPoint(0,0,0);
+
+  if( coord_type.compare("BrBz") == 0 ) {
```

Comment:
This is the single most time-critical function in this entire repository: it is evaluated for every step of every track in every event (the rest of this repository typically just runs on geometry initialization). Therefore I'm going to apply a lot more scrutiny to this, and I apologize in advance. Magnetic field lookups are often among the most CPU intensive aspects of our simulations.

- We should avoid rerunning this string comparison for every step: it should be done once, and we should use a boolean that is a trivial lookup.
- We should not create dynamically allocated `indices` and `deltas`, and `getIndices` should be on statically allocated variables too.
- There should be no need for a std::count over `indices`.
- Avoid the loop over `comp` in the 3D case.

Normally I'd encourage the use of transitions to std::vector. Just not here...

Reply:
I've tried benchmarking simulation performance for this. I get:

before
```
        Command being timed: "ddsim --runType batch --filter.tracker edep0 -v WARNING --numberOfEvents 100000 --compactFile ...
        User time (seconds): 5558.81
        System time (seconds): 1.14
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:32:41
```

after
```
        Command being timed: "ddsim --runType batch --filter.tracker edep0 -v WARNING --numberOfEvents 100000 --compactFile ...
        User time (seconds): 8117.60
        System time (seconds): 1.02
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:15:19
```

This is not great, I suppose.

Reply:
Ok, let me see what I can do to streamline the fieldComponents() function.

Reply:
I'm trying to understand which part of FieldMapB.cpp is causing the main slow down.
Most particles don't go all the way down to the FB area where the new lumi dipole field map exists.
Will a particle a mid rapidity call on the fieldComponents() for the FB area or just in the area that it traverses?

Reply:
@veprbl , thanks for checking the simulation performance time.  Can you check it again after my latest commit.  Hopefully it has been improved.

Reply:
Seems like we are back to normal:
```
        Command being timed: "ddsim --runType batch --filter.tracker edep0 -v WARNING --numberOfEvents 100000 --compactFile ...
        User time (seconds): 5715.81
        System time (seconds): 1.32
        Percent of CPU this job got: 89%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:46:54
```
Note there was some extra load on the machine, so look at the user time

Reply:
One more item to possibly optimize:
https://github.com/eic/epic/blob/9f53900013851663002bd88f096107e58508748a/src/FieldMapB.cpp#L231
It was mentioned that the field map could be transformed once during the load, and then remove this from critical fieldComponents() function.  If we only consider translations of the field map only, then it's simple.  However, things can be tricky with rotation transformations.  What is to be rotated?  The 3D position coordinates, or the field, or both?

Reply:
I don't see a simple way to move the field transformation to the load stage while still keeping the same number of calculations in GetIndices().
I suppose that this (3x3) x (3x1) multiplication is optimized automatically by the compiler.  So maybe it's ok to leave this coordinate transformation here as it is?

Reply:
@veprbl , @wdconinc , can we resolve this discussion of performance for this PR?  From the recent benchmark test that Dima just ran (thanks), it appears that there is no slow down with the changes.